samples/bluetooth/iso_broadcast_benchmark/src/receiver.c

```diff
@@ -0,0 +1,495 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <ctype.h>
+#include <device.h>
+#include <devicetree.h>
+#include <drivers/gpio.h>
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/iso.h>
+#include <sys/byteorder.h>
+#include <console/console.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(iso_broadcast_receiver, LOG_LEVEL_DBG);
+
+#define DEVICE_NAME	CONFIG_BT_DEVICE_NAME
+#define DEVICE_NAME_LEN (sizeof(DEVICE_NAME))
+
+#define BT_INTERVAL_TO_MS(interval) ((interval) * 5 / 4)
+#define PA_RETRY_COUNT 6
+#define ISO_RETRY_COUNT 10
+
+struct iso_recv_stats {
+	uint32_t     iso_recv_count;
+	uint32_t     iso_lost_count;
+};
+
+static bool         broadcaster_found;
+static bool         per_adv_lost;
+static bool         big_sync_lost;
+static bool         biginfo_received;
+static bt_addr_le_t per_addr;
+static uint8_t      per_sid;
+static uint16_t     per_interval_ms;
+static uint16_t     iso_interval_ms;
+static uint8_t      bis_count;
+static uint32_t     last_received_counter;
+static int64_t      big_sync_start_time;
+static size_t       big_sync_count;
+
+static struct iso_recv_stats stats_current_sync;
+static struct iso_recv_stats stats_overall;
+
+static K_SEM_DEFINE(sem_per_adv, 0, 1);
+static K_SEM_DEFINE(sem_per_sync, 0, 1);
+static K_SEM_DEFINE(sem_per_sync_lost, 0, 1);
+static K_SEM_DEFINE(sem_per_big_info, 0, 1);
+static K_SEM_DEFINE(sem_big_sync, 0, 1);
+static K_SEM_DEFINE(sem_big_sync_lost, 0, 1);
+
+static const char *phy2str(uint8_t phy)
+{
+	switch (phy) {
+	case 0: return "No packets";
+	case BT_GAP_LE_PHY_1M: return "LE 1M";
+	case BT_GAP_LE_PHY_2M: return "LE 2M";
+	case BT_GAP_LE_PHY_CODED: return "LE Coded";
+	default: return "Unknown";
+	}
+}
+
+static bool data_cb(struct bt_data *data, void *user_data)
+{
+	char *name = user_data;
+	uint8_t len;
+
+	switch (data->type) {
+	case BT_DATA_NAME_SHORTENED:
+	case BT_DATA_NAME_COMPLETE:
+		len = MIN(data->data_len, DEVICE_NAME_LEN - 1);
+		memcpy(name, data->data, len);
+		name[len] = '\0';
+		return false;
+	default:
+		return true;
+	}
+}
+
+static void scan_recv(const struct bt_le_scan_recv_info *info,
+		      struct net_buf_simple *buf)
+{
+	char le_addr[BT_ADDR_LE_STR_LEN];
+	char name[DEVICE_NAME_LEN];
+
+	if (broadcaster_found) {
+		return;
+	}
+
+	(void)memset(name, 0, sizeof(name));
+
+	bt_data_parse(buf, data_cb, name);
+
+	if (strncmp(DEVICE_NAME, name, strlen(DEVICE_NAME))) {
+		return;
+	}
+
+	bt_addr_le_to_str(info->addr, le_addr, sizeof(le_addr));
+
+	LOG_INF("Found broadcaster with address %s (RSSI %i)",
+		le_addr, info->rssi);
+
+	broadcaster_found = true;
+
+	per_sid = info->sid;
+	per_interval_ms = BT_INTERVAL_TO_MS(info->interval);
+	bt_addr_le_copy(&per_addr, info->addr);
+
+	k_sem_give(&sem_per_adv);
+}
+
+static struct bt_le_scan_cb scan_callbacks = {
+	.recv = scan_recv,
+};
+
+static void sync_cb(struct bt_le_per_adv_sync *sync,
+		    struct bt_le_per_adv_sync_synced_info *info)
+{
+	LOG_INF("Periodic advertisement synced");
+
```

Comment:
Should per_adv_lost be set to zero here (as well), in case of multiple runs/restarts?

Reply:
`reset_sems` ensures that all semaphores are reset when rerunning.

Reply:
per_adv_lost is a boolean, not a semaphore.
But I see now that this will be reset when test_run_receiver is run, and there are not other "internal" re-runs, so this is fine.